messaging/src/main/java/org/axonframework/common/annotation/AnnotationUtils.java

```diff
@@ -179,8 +270,51 @@ private static String resolveName(Method method) {
         return method.getName();
     }
 
-    private AnnotationUtils() {
-        // utility class
+    /**
+     * Validate whether the given {@code target} annotation {@link Class} is meta-annotated with the given {@code
+     * subject}. If this is the case for the {@code target} itself or any meta-annotation on any level of the {@code
+     * target}, {@code true} will be returned.
+     * <p>
+     * Any {@link Annotation} classes which are directly annotated or meta-annotated with the given {@code subject} will
+     * be stored in the {@code annotatedWithSubject} {@link Set}. The {@code visited} {@code Set} is used to ignore
+     * annotations which have already been validated.
+     *
+     * @param target               the annotation {@link Class} to validate if it is annotated with the given {@code
+     *                             subject}
+     * @param subject              the annotation {@link Class} to check whether it is present on the given {@code
+     *                             target}, directly or through meta-annotations
+     * @param annotatedWithSubject a {@link Set} to store all class' in which are annotated with the {@code subject},
+     *                             either directly or through meta-annotations
+     * @param visited              a {@link Set} containing all annotation class' which have been visited in the process
+     *                             to overcome an endless validation loop
+     * @return {@code true} if the {@code target} or any meta-annotations of the {@code target} are annotated with the
+     * {@code subject}, {@code false} otherwise
+     */
+    public static boolean isAnnotatedWith(Class<? extends Annotation> target,
```

Comment:
Might be worth splitting this method into 2 or 3 specialised methods. 

- Looking up and gathering of targets annotated with the subject can be a separate method. As this is a recursive method with state passed trough, it might be better to hide this fact (`visited` and `annotatedWithSubject` parameters) from the outside.
- `isAnnotatedWith` implies a boolean return value to answer if the target is annotated with the subject. This can use the method that will gather all targets annotated with the subject from the above point to do the complex work. If performance is critical, it can have its own search but short-circuit the moment the annotation is found

Reply:
Although yes, I can create an `isAnnotatedWith` which does not expose `visited` (the second method) and/or `annotatedWithSubject` (the third method), there is nothing using at all at this stage. I was in doubt about doing it anyway, but it felt like premature optimization to me as long as _nothing_ will be using that function at all.

If you feel strongly it clarifies things for our (future) selves and users, I'd gladly add it too. At any point, let me know. :-)

Reply:
I'm with it either way for now as long it's only used in one place. When we expand the usage to another place, it will make way more sense to at least remove mutability